### ISSUE TITLE
CH5-Logistic regression

### ISSUE THREAD

**Issue by Pin-Jiun** [Pin-Jiun/Machine-Learning-MIT#11]
### A new hypothesis class: linear logistic classifiers

Even for simple linear classifiers, it is very difficult to find values for θ, θ0 that minimize simple training error
![image](https://user-images.githubusercontent.com/34474924/233271715-be8b3b09-70f8-4584-89f9-5a99ddcff20f.png)

This problem is NP-hard, which probably implies that solving the most difficult instances of this problem would require computation time exponential in the number of training examples, n.

What makes this a difficult optimization problem is its lack of “smoothness”, 

Loss Function 不夠smoothness的原因
- 沒有辦法描述這個點對這個classifier有多糟糕, 不是+1就是-1
- 有可能有兩個hypotheses, 雖然這兩個都有相同的分類, 但是有可能有一個才是理想的classifier

For these reasons, we are going to investigate a new hypothesis class: linear logistic classifiers.

These hypotheses are still parameterized by a d-dimensional vector θ and a scalar θ0, but instead of making predictions in {+1,−1}, they generate real-valued outputs in the interval (0, 1). A linear logistic classifier has the form

![image](https://user-images.githubusercontent.com/34474924/233272739-b4e39cfa-b29f-4694-bbba-527cb07a8675.png)

The logistic function, also known as the sigmoid function, is defined as
![image](https://user-images.githubusercontent.com/34474924/233273053-db56b9da-354a-4a9f-bd1b-3604520892ec.png)

Its output can be interpreted as a probability, because for any value of z the output is in (0, 1)
![image](https://user-images.githubusercontent.com/34474924/233273228-b3284643-8d79-447d-b4d4-cc49b5eb73ef.png)

Given an LLC, with an output value in (0, 1), what should we do if we are forced to make a prediction in {+1,−1}?
A default answer is to predict +1 if σ(θT x + θ0) > 0.5 and −1 otherwise. 
The value 0.5 is sometimes called a `prediction threshold`.
for different problem settings, we might prefer to pick a different prediction threshold 
For example, if the consequences of predicting +1 when the answer should be −1 are much worse than the consequences of predicting −1 when the answer should be +1, then we might set the prediction threshold to be greater than 0.5.

When d = 2, then our inputs x lie in a two-dimensional space with axes x1 and x2, and the output of the LLC is a surface, as shown below, for θ = (1, 1), θ0 = 2.

![image](https://user-images.githubusercontent.com/34474924/233283282-35806a6b-ca45-4d5c-b799-39e099ce3d3f.png)

_______________

### negative log-likelihood (NLL) - log loss/cross entropy for logistic classifiers

sigmoid 的值介於0~1, 所以可以把它當作機率去解釋, 用這個特性去產生Loss function
![image](https://user-images.githubusercontent.com/34474924/233284656-0d1053e8-01c4-4a5c-afc3-da9d7c6052e3.png)

In order to simplify the description, we will assume that (or transform so that) the labels in the training data are y ∈ {0, 1}
under the assumption that our predictions are independent. This can be cleverly rewritten,
![image](https://user-images.githubusercontent.com/34474924/233286435-b1d17179-213f-4ab3-973f-ff1b470630f5.png)
機率越大代表分類的準確率越高, 但乘法很難算
這時使用log function, 因為log function is monotonic, 當x越大log(x)也越大,反之x越小log(x)也越小, 呈現正相關

![image](https://user-images.githubusercontent.com/34474924/233287037-7b29209c-7e0a-43ee-a99e-0ef9a344b39a.png)
但我們想要得到Loss function, 也就是loss越小越好的function, 所以取個負號就可輕鬆達成
![image](https://user-images.githubusercontent.com/34474924/233287408-2a3fad84-7021-4b51-8047-e3e503d3624b.png)
This loss function is also sometimes referred to as the log loss or cross entropy.

最後將Regularization帶入, 得到完整的Loss function
![image](https://user-images.githubusercontent.com/34474924/233289697-b7219024-d438-46d1-bf38-ba48a777200b.png)
